File: 100 - The 3n + 1 Problem/3n1.cpp
// This is a more typical example of the problem *without* an acceleration
// structure.

#include <iostream>
#include <string>

int recurse(int num);

int test(int low, int high);

int main() {
    bool cont = true;
    int first, second, temp, max = 0;
    std::size_t space = 0;

    while (cont == true) {
        std::string input = "";
        std::getline(std::cin, input);

        if (input.size() != 0) {
            space = 0;
            first = std::stoi(input, &space);
            second = std::stoi(input.substr(space));
            temp = 0;
            max = test(first, second);
            std::cout << first << " " << second << " " << max << "\n";
        } else {
            cont = false;
        }
    }
    return 0;
}

int recurse(int num) {
    int res = 0;
    int tick = 0;

    if (num % 2) {
        res = (3 * num) + 1;
    } else {
        res = num / 2;
    }

    if (res == 1) {
        tick = 1;
    } else {
        tick = recurse(res);
    }
    return tick + 1;
}

int test(int low, int high) {
    int test, max = 0;
    if (low > high) {
        test = low;
        low = high;
        high = test;
    }

    for (int i = low; i <= high; i++) {
        if (i == 1) {
            test = 1;
        } else {
            test = recurse(i);
        }
        if (test > max) {
            max = test;
        }
    }
    return max;
}

File: 100 - The 3n + 1 Problem/3n1.pas
program threenone(input, output);
uses
    sysUtils;
var
    i : longInt; // Control variable, controls loops
    first, second, temp, max : longInt;
    cont, lineEmpty : boolean; // Control variable, exits program
    space : SizeInt;
    line : string;

    function recurse (num : longInt) : longInt;
    var
        res : longInt = 0;
        tick : longInt = 0;
    begin
        if (num mod 2) > 0 then
            res := (3 * num) + 1
        else
            res := round(num / 2);

        if res = 1 then
            tick := 1
        else
            tick := recurse(res);
        
        recurse := tick + 1;
    end;

    function test (low : longInt; high : longInt) : longInt;
    var
        testVar : longInt = 0;
        max : longInt = 0;
    begin
        if low > high then
        begin
            testVar := low;
            low := high;
            high := testVar;
        end;

        for i := low to high do
        begin
            if i = 1 then
                testVar := 1
            else
                testVar := recurse(i);
            
            if testVar > max then
                max := testVar;
        end;

        test := max;
    end;

begin
    cont := true;
    lineEmpty := true;

    while cont = true do
    begin
        // line := '';qu
        readln(line);
        line := Trim(line);
        if Length(line) > 0 then
        begin
            while Length(line) > 0 do
            begin
                space := LastDelimiter(' ', line);
                second := StrToInt(RightStr(line, (Length(line)-space)));
                Delete(line, space, (Length(line)-space)+1);
                line := Trim(line);
                space := LastDelimiter(' ', line);
                if space > 0 then
                begin
                    first := StrToInt(RightStr(line, (Length(line)-space)));
                    Delete(line, space, (Length(line)-space)+1);
                    line := Trim(line);
                end
                else
                begin
                    first := StrToInt(line);
                    line := '';
                end;
                temp := 0;
                max := test(first, second);
                writeln(IntToStr(first) + ' ' + IntToStr(second) + ' ' + IntToStr(max))
            end;
        end
        else
            cont := false;
    end;
end.